File: Editorial/1123_Lowest_Common_Ancestor_of_Deepest_Leaves/JavaScript/1_Recursion.js
var lcaDeepestLeaves = function (root) {
    return dfs(root)[1];
};

function dfs(root) {
    if (!root) {
        return [0, null];
    }

    let left = dfs(root.left);
    let right = dfs(root.right);

    if (left[0] > right[0]) {
        return [left[0] + 1, left[1]];
    }
    if (left[0] < right[0]) {
        return [right[0] + 1, right[1]];
    }
    return [left[0] + 1, root];
}

File: Editorial/1101-1200/1123_Lowest_Common_Ancestor_of_Deepest_Leaves/CSharp/1_Recursion.cs
public class Solution {
    public TreeNode LcaDeepestLeaves(TreeNode root) {
        return dfs(root).Item1;
    }

    private Tuple<TreeNode, int> dfs(TreeNode root) {
        if (root == null) {
            return new Tuple<TreeNode, int>(null, 0);
        }

        Tuple<TreeNode, int> left = dfs(root.left);
        Tuple<TreeNode, int> right = dfs(root.right);

        if (left.Item2 > right.Item2) {
            return new Tuple<TreeNode, int>(left.Item1, left.Item2 + 1);
        }
        if (left.Item2 < right.Item2) {
            return new Tuple<TreeNode, int>(right.Item1, right.Item2 + 1);
        }
        return new Tuple<TreeNode, int>(root, left.Item2 + 1);
    }
}
